src/twisted/internet/test/test_endpoints.py

```diff
@@ -1687,6 +1688,412 @@ class WithResolver(proxyForInterface(
+
+        def assertHostPortFamilySockType(result):
+            self.assertEqual(len(result), 1)
+            [(family, socktype, _, _, sockaddr)] = result
+            self.assertEqual(family, AF_INET)
+            self.assertEqual(socktype, SOCK_STREAM)
+            self.assertEqual(sockaddr, (host, port))
+
+        return resolutionDeferred.addCallback(assertHostPortFamilySockType)
+
+
+
+class _HostnameEndpointMemoryReactorMixin(ClientEndpointTestCaseMixin):
+    """
+    Common methods for testing L{HostnameEndpoint} against
+    L{MemoryReactor} instances that do not provide
+    L{IReactorPluggableNameResolver}.
+    """
+
+    def synchronousDeferredToThread(self, f, *args, **kwargs):
+        """
+        A synchronous version of L{deferToThread}.
+
+        @param f: The callable to invoke.
+        @type f: L{callable}
+
+        @param args: Positional arguments to the callable.
+
+        @param kwargs: Keyword arguments to the callable.
+
+        @return: A L{Deferred} that fires with the result of applying
+            C{f} to C{args} and C{kwargs} or the exception raised.
+        """
+        try:
+            result = f(*args, **kwargs)
+        except:
+            return defer.fail()
+        else:
+            return defer.succeed(result)
+
+
+    def expectedClients(self, reactor):
+        """
+        Extract expected clients from the reactor.
+
+        @param reactor: The L{MemoryReactor} under test.
+
+        @return: List of calls to L{IReactorTCP.connectTCP}
+        """
+        return reactor.tcpClients
+
+
+    def connectArgs(self):
+        """
+
+        @return: C{dict} of keyword arguments to pass to connect.
+        """
+        return {'timeout': 10, 'bindAddress': ('localhost', 49595)}
+
+
+    def assertConnectArgs(self, receivedArgs, expectedArgs):
+        """
+        Compare host, port, timeout, and bindAddress in C{receivedArgs}
+        to C{expectedArgs}.  We ignore the factory because we don't
+        only care what protocol comes out of the
+        C{IStreamClientEndpoint.connect} call.
+
+        @param receivedArgs: C{tuple} of (C{host}, C{port}, C{factory},
+            C{timeout}, C{bindAddress}) that was passed to
+            L{IReactorTCP.connectTCP}.
+        @param expectedArgs: C{tuple} of (C{host}, C{port}, C{factory},
+            C{timeout}, C{bindAddress}) that we expect to have been passed
+            to L{IReactorTCP.connectTCP}.
+        """
+        (host, port, ignoredFactory, timeout, bindAddress) = receivedArgs
+        (expectedHost, expectedPort, _ignoredFactory,
+         expectedTimeout, expectedBindAddress) = expectedArgs
+
+        self.assertEqual(host, expectedHost)
+        self.assertEqual(port, expectedPort)
+        self.assertEqual(timeout, expectedTimeout)
+        self.assertEqual(bindAddress, expectedBindAddress)
+
+
+    def test_endpointConnectFailure(self):
+        """
+        If L{HostnameEndpoint.connect} is invoked and there is no server
```

Comment:
"when" is preferable to "if", because it stipulates the existence of a supported configuration.